packages/firestore/test/integration/util/composite_index_test_helper.ts

```diff
@@ -0,0 +1,165 @@
+  }
+
+  // Runs a test with specified documents in the COMPOSITE_INDEX_TEST_COLLECTION.
+  async withTestDocs<T>(
+    persistence: PersistenceMode | typeof PERSISTENCE_MODE_UNSPECIFIED,
+    docs: { [key: string]: DocumentData },
+    fn: (collection: CollectionReference, db: Firestore) => Promise<T>
+  ): Promise<T> {
+    return batchCommitDocsToCollection(
+      persistence,
+      DEFAULT_SETTINGS,
+      this.hashDocs(docs),
+      COMPOSITE_INDEX_TEST_COLLECTION,
+      fn
+    );
+  }
+
+  // Runs a test without pre-created documents in the COMPOSITE_INDEX_TEST_COLLECTION.
+  async withEmptyCollection<T>(
+    persistence: PersistenceMode | typeof PERSISTENCE_MODE_UNSPECIFIED,
+    fn: (collection: CollectionReference, db: Firestore) => Promise<T>
+  ): Promise<T> {
+    return batchCommitDocsToCollection(
+      persistence,
+      DEFAULT_SETTINGS,
+      {},
+      COMPOSITE_INDEX_TEST_COLLECTION,
+      fn
+    );
+  }
+
+  // Hash the document key and add testId to documents created under a specific test to support data
+  // isolation in parallel testing.
+  private hashDocs(docs: { [key: string]: DocumentData }): {
+    [key: string]: DocumentData;
+  } {
+    return Object.keys(docs).reduce((result, key) => {
+      const doc = { ...docs[key], [this.TEST_ID_FIELD]: this.testId };
+      result[key + '-' + this.testId] = doc;
+      return result;
+    }, {} as { [key: string]: DocumentData });
+  }
+
+  // Hash the document keys with testId.
+  toHashedIds(docs: string[]): string[] {
+    return docs.map(docId => docId + '-' + this.testId);
+  }
+
+  // Checks that running the query while online (against the backend/emulator) results in the same
+  // as running it while offline. The expected document Ids are hashed to match the actual document
+  // IDs created by the test helper.
+  async checkOnlineAndOfflineResultsMatch(
+    query: Query,
+    ...expectedDocs: string[]
+  ): Promise<void> {
+    return checkOnlineAndOfflineResultsMatch(
+      query,
+      ...this.toHashedIds(expectedDocs)
+    );
+  }
+
+  // Adds a filter on test id for a query.
+  query<AppModelType, DbModelType extends DocumentData>(
+    query_: Query<AppModelType, DbModelType>,
+    ...queryConstraints: QueryConstraint[]
+  ): Query<AppModelType, DbModelType> {
+    return internalQuery(
+      query_,
+      where(this.TEST_ID_FIELD, '==', this.testId),
+      ...queryConstraints
+    );
+  }
+
+  // Adds a filter on test id for a composite query,
+  compositeQuery<AppModelType, DbModelType extends DocumentData>(
+    query_: Query<AppModelType, DbModelType>,
+    compositeFilter: QueryCompositeFilterConstraint,
+    ...queryConstraints: QueryNonFilterConstraint[]
+  ): Query<AppModelType, DbModelType> {
+    return internalQuery(
+      query_,
+      and(where(this.TEST_ID_FIELD, '==', this.testId), compositeFilter),
+      ...queryConstraints
+    );
+  }
+
+  // Add a document with test id.
+  addDoc<T, DbModelType extends DocumentData>(
```

Comment:
addDoc and setDoc is not used in current test cases. Should I keep these for future use or remove it as it is not required for now?